chain-signatures/node/src/indexer_eth.rs

```diff
@@ -463,55 +719,162 @@ async fn process_block(
-                    .as_secs() as f64,
-            );
+    let mut final_block_number: Option<u64> = None;
+    let mut finalized_epoch: BlockNumberToHashMap = HashMap::new();
+    while let Some(BlockAndRequests {
+        block_number,
+        block_hash,
+        indexed_requests,
+    }) = requests_indexed.recv().await
+    {
+        // keep looping until the most recent updated finalized epoch contains the block number
+        loop {
+            if let Some(finalized_block_hash) = finalized_epoch.get(&block_number) {
+                if *finalized_block_hash == block_hash {
+                    tracing::info!("Block {block_number} is finalized!");
+                    send_indexed_requests(
+                        indexed_requests,
+                        sign_tx.clone(),
+                        node_near_account_id.clone(),
+                    )
+                    .await;
+                    break;
+                } else {
+                    tracing::warn!(
+                                "Block {block_number} hash mismatch: expected {block_hash:?}, got {finalized_block_hash:?}. Chain re-orged."
+                            );
+                    //TODO: handle the block reorg case
+                    break;
+                }
+            } else {
+                tracing::warn!(
+                    "Block number {block_number} with hash: {block_hash:?} not in finalized epoch. "
+                );
+                if !finalized_epoch.is_empty() && finalized_epoch.keys().all(|&k| k > block_number)
+                {
+                    tracing::warn!("Block {block_number} is in history");
+                    //TODO: handle the block in history case which happens when a block is retried
+                    continue;
+                } else {
+                    // cur request is not finalized yet, poll for the next finalized block.
+                    // Break the loop when we get an update on final block number and finalized epoch.
+                    loop {
+                        let Ok((new_final_block_number, new_finalized_epoch)) =
+                            refresh_finalized_epoch(
+                                helios_client,
+                                untrusted_rpc_client,
+                                final_block_number,
+                                refresh_finalized_interval,
+                            )
+                            .await
+                        else {
+                            tracing::warn!("Failed to refresh finalized epoch");
+                            continue;
+                        };
+                        if let Some(new_final_block_number) = new_final_block_number {
+                            final_block_number.replace(new_final_block_number);
+                            finalized_epoch.clear();
+                            finalized_epoch.extend(new_finalized_epoch);
```

Comment:
This code can be removed if you pass `&mut finalized_epoch` to `refresh_finalized_epoch`. Same for `final_block_number`.

It's a matter of style. Some people really don't like &mut arguments. But IMO, this is a case where the variable is read inside and we do factually update it. It's just that now you do in an extra step that always happens afterwards.